promql/test.go

```diff
@@ -416,6 +432,157 @@ func (cmd *loadCmd) append(a storage.Appender) error {
 			}
 		}
 	}
+	if cmd.withNhcb {
+		return cmd.appendCustomHistogram(a)
+	}
+	return nil
+}
+
+func getHistogramMetricBase(m labels.Labels, suffix string) (labels.Labels, uint64) {
+	mName := m.Get(labels.MetricName)
+	baseM := labels.NewBuilder(m).
+		Set(labels.MetricName, strings.TrimSuffix(mName, suffix)).
+		Del(labels.BucketLabel).
+		Labels()
+	hash := baseM.Hash()
+	return baseM, hash
+}
+
+// if classic histograms are defined, convert them into native histograms with custom
+// bounds and append the defined time series to the storage.
+func (cmd *loadCmd) appendCustomHistogram(a storage.Appender) error {
```

Comment:
This algorithm is pretty complicated.
I'm thinking that this could be simpler if we kept a map[uint64]map[float]*histogram.Histogram. 
Where we map from hash of series (without suffix or `le` label)  to   map of timestamp to the resulting histogram.


So when you hit a _count or _sum series, you remove the suffix, take the hash and find/create the histogram at the timestamp and set count or sum.
When you hit a _bucket, you'd again remove the suffix and `le` label and find/create the histogram at the timestamp and *insert* the bound as appropriate at `le` value into CustomValues (incl. +Inf for now)  and also the bucket count into PositiveBuckets as an absolute value at the same position as CustomValues.

When all is done, you take another pass and turn the absolute values into deltas in PositiveBuckets and remove +Inf from the custom bounds. And also check the overall count and do any other check.

Reply:
I see what you mean about it being messy right now, but I feel that continually inserting bounds and counts and updating the spans in the histogram is also kind of messy, and while it's going through this process the histogram would be very invalid, i.e. total count and bucket counts will be wrong since you need to make it uncumulative afterwards, and you'd have an +Inf bound that isn't supposed to be there. How about I create a temporary struct just for this instead, then the second pass can convert that temp struct into a histogram and check validity?

Reply:
I think a temporary struct is good compromise.